src/Features/Core/Portable/PopulateSwitch/AbstractPopulateSwitchDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,121 @@
+
+        private static readonly DiagnosticDescriptor s_descriptor = new DiagnosticDescriptor(IDEDiagnosticIds.PopulateSwitchDiagnosticId,
+                                                                    s_localizableTitle,
+                                                                    s_localizableMessage,
+                                                                    DiagnosticCategory.EditAndContinue,
+                                                                    DiagnosticSeverity.Warning,
+                                                                    isEnabledByDefault: true,
+                                                                    customTags: DiagnosticCustomTags.Unnecessary);
+
+        #region Interface methods
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_descriptor);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(
+                nodeContext =>
+                {
+                    Diagnostic diagnostic;
+                    if (TryPopulateSwitch(nodeContext.SemanticModel, nodeContext.Node, out diagnostic, nodeContext.CancellationToken))
+                    {
+                        nodeContext.ReportDiagnostic(diagnostic);
+                    }
+                },
+                SyntaxKindsOfInterest);
+        }
+
+        protected abstract ImmutableArray<TLanguageKindEnum> SyntaxKindsOfInterest { get; }
+
+        #endregion
+
+        protected abstract SyntaxNode GetExpression(SyntaxNode node);
+        protected abstract List<SyntaxNode> GetCaseLabels(SyntaxNode node, out bool hasDefaultCase);
+
+        private bool SwitchIsFullyPopulated(SemanticModel model, SyntaxNode node)
+        {
+            var enumType = model.GetTypeInfo(GetExpression(node)).Type as INamedTypeSymbol;
+            if (enumType == null || enumType.TypeKind != TypeKind.Enum)
+            {
+                return true;
+            }
+
+            // ignore enums marked with Flags per spec: https://github.com/dotnet/roslyn/issues/6766#issuecomment-156878851
+            foreach (var attribute in enumType.GetAttributes())
+            {
+                var containingClass = attribute.AttributeClass.ToDisplayString();
+                if (containingClass == typeof(System.FlagsAttribute).FullName)
+                {
+                    return true;
+                }
+            }
+
+            bool hasDefaultCase;
+            var caseLabels = GetCaseLabels(node, out hasDefaultCase);
+
+            if (!hasDefaultCase)
+            {
+                return false;
+            }
+
+            var labelSymbols = new List<ISymbol>();
+            foreach (var label in caseLabels)
+            {
+                labelSymbols.Add(model.GetSymbolInfo(label).Symbol);
```

Comment:
I have never had it happen in my tests, but I should probably create a null check just in case.  If a symbol returns null, should I abort the check as the code is probably uncompilable anyway?